src/main/scala/mesosphere/marathon/upgrade/DeploymentPlan.scala

```diff
@@ -321,7 +321,65 @@ object DeploymentPlan {
 
     validator[DeploymentPlan] { plan =>
       plan.createdOrUpdatedApps as "app" is every(valid(AppDefinition.updateIsValid(plan.original)))
-      plan should notBeTooBig
+      plan should notBeTooBig and haveValidPorts
+    }
+  }
+
+  private def haveValidPorts: Validator[DeploymentPlan] = {
+    new Validator[DeploymentPlan] {
+      override def apply(plan: DeploymentPlan): Result = {
+        // Each service port should be unique across a cluster. We
+        // want to report conflicts if the same service port is used
+        // by multiple apps.
+
+        // This validation rule is only applied for newly created or
+        // updated applications (within this deployment plan),
+        // not for all applications.
+
+        // We construct the mapping servicePort <- Set[AppKey]. This
+        // allows us to report any service port that has more than 1
+        // application using the same port.
+
+        // We keep track of the total number of ports to support an
+        // early exit condition.
+
+        val createdOrUpdatedApps = plan.createdOrUpdatedApps.map(app => app.id).toSet
+        var ports: Int = 0
+        val merged =
+          new mutable.HashMap[Int, mutable.Set[AppDefinition.AppKey]] with mutable.MultiMap[Int, AppDefinition.AppKey]
+
+        // Add each servicePort <- Application to the map.
+        for {
+          app <- plan.target.transitiveApps
+          // We ignore randomly assigned ports identified by `0`.
+          port <- app.servicePorts if port != 0
+        } {
+          ports += 1
+          merged.addBinding(port, app.id)
+        }
+
+        // If the total number of unique ports is equal to the number
+        // of requested ports then we know there are no conflicts.
+        if (merged.size == ports) {
+          Success
+        } else {
+          // Otherwise we find all ports that have more than 1 app
+          // interested in them.
+
+          // We report all the conflicting apps along with which other
+          // apps they conflict with.
+          val violations = for {
+            // only yield port as conflict if createdOrUpdatedApps are involved and therefore in the intersection set
+            (port, apps) <- merged if apps.size > 1 && apps.intersect(createdOrUpdatedApps).nonEmpty
```

Comment:
This will filter the list of `(port, apps)`. It could lead to an empty array for violations, which still would cause a Failure - or (the filter is applied in the else/failure branch)?